core/src/main/java/org/elasticsearch/index/query/ScriptQueryBuilder.java

```diff
@@ -158,17 +158,23 @@ public String toString(String field) {
 
         @Override
         public boolean equals(Object obj) {
-            if (this == obj)
+            // TODO: Do this if/when we can assume scripts are pure functions
+            // and they have a reliable equals impl
+            /*if (this == obj)
                 return true;
             if (sameClassAs(obj) == false)
                 return false;
             ScriptQuery other = (ScriptQuery) obj;
-            return Objects.equals(script, other.script);
+            return Objects.equals(script, other.script);*/
+            return false;
```

Comment:
Caching could still happen in that case if you use the same weight create a Scorer on the same segment multiple times.